File: engine/core/io/UserSettings.cpp
//
// (c) 2020 Eduardo Doria.
//

#include "UserSettings.h"
#include "System.h"

using namespace Supernova;

bool UserSettings::getBoolForKey(const char *key){
    return getBoolForKey(key, false);
}

int UserSettings::getIntegerForKey(const char *key){
    return getIntegerForKey(key, 0);
}

long UserSettings::getLongForKey(const char *key){
    return getLongForKey(key, 0);
}

float UserSettings::getFloatForKey(const char *key){
    return getFloatForKey(key, 0);
}

double UserSettings::getDoubleForKey(const char *key){
    return getDoubleForKey(key, 0);
}

std::string UserSettings::getStringForKey(const char *key){
    return getStringForKey(key, "");
}

Data UserSettings::getDataForKey(const char *key){
    return getDataForKey(key, Data());
}

bool UserSettings::getBoolForKey(const char *key, bool defaultValue){
    return System::instance().getBoolForKey(key, defaultValue);
}

int UserSettings::getIntegerForKey(const char *key, int defaultValue){
    return System::instance().getIntegerForKey(key, defaultValue);
}

long UserSettings::getLongForKey(const char *key, long defaultValue){
    return System::instance().getLongForKey(key, defaultValue);
}

float UserSettings::getFloatForKey(const char *key, float defaultValue){
    return System::instance().getFloatForKey(key, defaultValue);
}

double UserSettings::getDoubleForKey(const char *key, double defaultValue){
    return System::instance().getDoubleForKey(key, defaultValue);
}

std::string UserSettings::getStringForKey(const char *key, std::string defaultValue){
    return System::instance().getStringForKey(key, defaultValue);
}

Data UserSettings::getDataForKey(const char *key, const Data& defaultValue){
    return System::instance().getDataForKey(key, defaultValue);
}

void UserSettings::setBoolForKey(const char *key, bool value){
    System::instance().setBoolForKey(key, value);
}

void UserSettings::setIntegerForKey(const char *key, int value){
    System::instance().setIntegerForKey(key, value);
}

void UserSettings::setLongForKey(const char *key, long value){
    System::instance().setLongForKey(key, value);
}

void UserSettings::setFloatForKey(const char *key, float value){
    System::instance().setFloatForKey(key, value);
}

void UserSettings::setDoubleForKey(const char *key, double value){
    System::instance().setDoubleForKey(key, value);
}

void UserSettings::setStringForKey(const char* key, std::string value){
    System::instance().setStringForKey(key, value);
}

void UserSettings::setDataForKey(const char *key, Data& value){
    System::instance().setDataForKey(key, value);
}

void UserSettings::removeKey(const char *key){
    System::instance().removeKey(key);
}

File: platform/android/java/org/supernovaengine/supernova/UserSettings.java
package org.supernovaengine.supernova;

import android.app.Activity;
import android.content.Context;
import android.content.SharedPreferences;

public class UserSettings {
    private Activity activity;

    private static final String PREFS_FILE_NAME = "UserSettings";

    public UserSettings(final Activity activity) {
        this.activity = activity;
    }

    public boolean getBoolForKey(String key, boolean defaultValue) {
        SharedPreferences sharedPref = activity.getSharedPreferences(PREFS_FILE_NAME, Context.MODE_PRIVATE);

        try {
            return sharedPref.getBoolean(key, defaultValue);
        } catch (Exception ex) {
            ex.printStackTrace();
        }

        return defaultValue;
    }

    public int getIntegerForKey(String key, int defaultValue) {
        SharedPreferences sharedPref = activity.getSharedPreferences(PREFS_FILE_NAME, Context.MODE_PRIVATE);

        try {
            return sharedPref.getInt(key, defaultValue);
        } catch (Exception ex) {
            ex.printStackTrace();
        }

        return defaultValue;
    }

    public long getLongForKey(String key, long defaultValue) {
        SharedPreferences sharedPref = activity.getSharedPreferences(PREFS_FILE_NAME, Context.MODE_PRIVATE);

        try {
            return sharedPref.getLong(key, defaultValue);
        } catch (Exception ex) {
            ex.printStackTrace();
        }

        return defaultValue;
    }

    public float getFloatForKey(String key, float defaultValue) {
        SharedPreferences sharedPref = activity.getSharedPreferences(PREFS_FILE_NAME, Context.MODE_PRIVATE);

        try {
            return sharedPref.getFloat(key, defaultValue);
        } catch (Exception ex) {
            ex.printStackTrace();
        }

        return defaultValue;
    }

    public double getDoubleForKey(String key, double defaultValue) {
        // SharedPreferences doesn't support double, using long to store
        SharedPreferences sharedPref = activity.getSharedPreferences(PREFS_FILE_NAME, Context.MODE_PRIVATE);

        try {
            return Double.longBitsToDouble(sharedPref.getLong(key, Double.doubleToRawLongBits(defaultValue)));
        } catch (Exception ex) {
            ex.printStackTrace();
        }

        return defaultValue;
    }

    public String getStringForKey(String key, String defaultValue) {
        SharedPreferences sharedPref = activity.getSharedPreferences(PREFS_FILE_NAME, Context.MODE_PRIVATE);

        try {
            return sharedPref.getString(key, defaultValue);
        } catch (Exception ex) {
            ex.printStackTrace();
        }

        return defaultValue;
    }

    public void setBoolForKey(String key, boolean value) {
        SharedPreferences sharedPref = activity.getSharedPreferences(PREFS_FILE_NAME, Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = sharedPref.edit();
        editor.putBoolean(key, value);
        editor.apply();
    }

    public void setIntegerForKey(String key, int value) {
        SharedPreferences sharedPref = activity.getSharedPreferences(PREFS_FILE_NAME, Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = sharedPref.edit();
        editor.putInt(key, value);
        editor.apply();
    }

    public void setLongForKey(String key, long value) {
        SharedPreferences sharedPref = activity.getSharedPreferences(PREFS_FILE_NAME, Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = sharedPref.edit();
        editor.putLong(key, value);
        editor.apply();
    }

    public void setFloatForKey(String key, float value) {
        SharedPreferences sharedPref = activity.getSharedPreferences(PREFS_FILE_NAME, Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = sharedPref.edit();
        editor.putFloat(key, value);
        editor.apply();
    }

    public void setDoubleForKey(String key, double value) {
        // SharedPreferences doesn't support double, using long to store
        SharedPreferences sharedPref = activity.getSharedPreferences(PREFS_FILE_NAME, Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = sharedPref.edit();
        editor.putLong(key, Double.doubleToRawLongBits(value));
        editor.apply();
    }

    public void setStringForKey(String key, String value) {
        SharedPreferences sharedPref = activity.getSharedPreferences(PREFS_FILE_NAME, Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = sharedPref.edit();
        editor.putString(key, value);
        editor.apply();
    }

    public void removeKey(String key) {
        SharedPreferences sharedPref = activity.getSharedPreferences(PREFS_FILE_NAME, Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = sharedPref.edit();
        editor.remove(key);
        editor.apply();
    }
}
